### DISCUSSION TITLE
When using on-demand ISR and i18n routing, localized paths must be revalidated manually?

### DISCUSSION THREAD

**Discussion by sumul** [vercel/next.js#40489] [Help]
I'm using both on-demand ISR and internationalized routing (with sub-paths) for https://figure.game/. On-demand ISR is how I deploy new daily puzzle data at midnight UTC every day, which works beautifully. The built-in i18n routing is working perfectly as well.

I just ran into a **little gotcha when using both** together: the localized versions of my routes weren't getting revalidated (so no new puzzle today for Spanish-speaking players). I think I've sorted it out, but wanted to get some feedback on my approach and also share this in case it's helpful for anyone else who runs into this issue. 

From the documentation on [i18n routing](https://nextjs.org/docs/advanced-features/i18n-routing#dynamic-routes-and-getstaticprops-pages), I had assumed that `revalidate()` would update `getStaticProps` for all locales. I think this bit, which seems to only apply to the initial build, is where I made this false assumption: 
> For [Automatically Statically Optimized](https://nextjs.org/docs/advanced-features/automatic-static-optimization) and non-dynamic getStaticProps pages, a version of the page will be generated for each locale.

When I double-checked the [on-demand ISR documentation](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration#using-on-demand-revalidation), this comment in the example code clued me in:
>     // this should be the actual path not a rewritten path
>     // e.g. for "/blog/[slug]" this should be "/blog/post-1"

I believe the solution is iterate over the i18n locales when revalidating paths: 

```javascript
// /pages/api/isr-api-route.js
import { i18n } from "../../next.config";
const REVALIDATE_PATHS = ["/", "/congrats"];

export default async function handler(req, res) {
  // ...

  // Make path components out of each i18n locale (plus blank for default)
  const locales = ["", ...i18n.locales.map((l) => `/${l}`)];

  // Revalidate each path and its localized version
  for await (const locale of locales) {
    for await (const path of REVALIDATE_PATHS) {
      await res.revalidate(locale + path);
    }
  }

  // ...
}
```

Does this seem like the right approach, or is there a better way?

**Comment by bhaveyx** [comment]
Any update on this?

**Comment by aleksakojadinovic** [comment]
Apparently when using domain-based routing the problem is even worse. It seems impossible to revalidate routes other than lose for default locale.

**Reply by shinwonse** [reply]
Is there any solution for using domain-based routing?

**Reply by aleksakojadinovic** [reply]
Sorry for not updating earlier. I've figured it out in the end - even though you use domain-based routing, the paths need to have language subpaths when revalidating manually. Therefore I ended up doing the same thing as suggested in the discussion itself, something like this:
```js
const INTL_PREFIXES = ['en', 'de', 'ru', ...whatever];

// URLS is an array of paths you wish to revalidate
const fullUrls = URLS
    .map((baseUrl) => INTL_PREFIXES.map((prefix) => urlJoin('/', prefix, baseUrl)))
    .flat();

await Promise.all(fullUrls.map((URL) => res.revalidate(URL)));
```

**Reply by shinwonse** [reply]
Thank you for your reply. My domain looks like 'kr.xxx.com', 'cn.xxx.com', 'jp.xxx.com', 'www.xxx.com'. Is it okay with your solution?

**Reply by aleksakojadinovic** [reply]
Yeah it should work with subdomains as well, a few of my locales are also subdomain-routed.

The way I figured it out is by browsing the `.next` build folder and checking the contents of `.next/server/pages` directory. If you have a folder for each of your supported locales there then it should work, because apparently the revalidation paths match this folder structure rather than the actual URL, which is a little bizzare if you ask me, but oh well :)

**Reply by shinwonse** [reply]
Thank you for your kind explanation. But I have a one question. `res.revalidate` get one parameter which name is urlPath. Is it okay give full url to `res.revalidate`?

```typescript
import type { NextApiRequest, NextApiResponse } from 'next';

import nextConfig from '../../../next.config';

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
  if (req.query.secret !== 'notion') {
    return res.status(401).json({ message: 'Invalid token' });
  }

  const domains = nextConfig.i18n.domains;
  const urls: string[] = domains.map((domain: { defaultLocale: string; domain: string }) => {
    const baseUrl = domain.domain;
    const path = `/privacy-policy/${req.query.slug}`;
    return `${baseUrl}${path}`;
  });

  try {
    await Promise.all(urls.map(url => res.revalidate(url)));
    return res.json({ revalidated: true });
  } catch (err) {
    return res.status(500).json({ message: 'Error revalidating' });
  }
}
```
This is my code, thanks!!

**Reply by aleksakojadinovic** [reply]
I've never tried it but it probably won't work with full url such as `mysite.com/en/some/path` but rather `/en/some/path`. Therefore instead of using 
```js
const baseUrl = domain.domain;
// ...
```
I think it should be
```js
const baseUrl = domain.defaultLocale; // <--- Just the locale prefix, not the full URL
const path = `/privacy-policy/${req.query.slug}`;
return `/${baseUrl}${path}`;
```


Notice that even though your actual URL could be `es.mysite.com/some/path`, you still need to specify the revalidation path like `/es/some/path`. This is because within the request itself Next.js is not aware of your actual domain, this is handled by your proxy depending on your production setup, so you should keep your paths relative.

**Reply by shinwonse** [reply]
@aleksakojadinovic Thank you!!!!!!! I finally solved it with your solution. You saved my time 🙏

**Reply by aleksakojadinovic** [reply]
No problem, glad to help :)

**Comment by edward-meister** [comment]
hmmm
my problem is that, because of rewrites, my localized paths are not even cached. 
i am using nextjs 15 and next-intl
any idea how to solve this?
